### ISSUE TITLE
TFLite GPU Delegate has problem with MobileNetV2

### ISSUE THREAD

**Issue by ramtin2080** [tensorflow/tensorflow#25131]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution : Linux Ubuntu 16.04
- Mobile device : Samsung Galaxy S9 Android 8.0 - Nexus 10 Android 5.1
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): b'v1.12.0-rc2-3-ga6d8ffae09' 1.12.0 (0.0.0-gpu-experimental for mobile device)
- Python version: 3.6

**Describe the current behavior**
GPU delegate has problem with MobileNetV2. When I select GPU from device list at tflite demo project (https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/java/demo) the app crashes.
The only things I've changed on this project was changing MobileNet V1 float model to MobileNet v2. The MobileNetV2 model is taken from "https://tfhub.dev/google/imagenet/mobilenet_v2_050_160/classification/2" , retrained by "https://github.com/tensorflow/hub/blob/master/examples/image_retraining/retrain.py" script and converted to tflite format using the following command:

```
tflite_convert \
  --output_file=graph.tflite \
  --graph_def_file=retrained_graph.pb \
  --input_arrays=Placeholder \
  --output_arrays=final_result
 --input_shapes=1,160,160,3
```
All the necessary changes (such as changing the graph name, input size, etc.) in the ImageClassifierFloatMobileNet class is made.

Logs:
```
2019-01-23 13:51:12.091 22222-22294/android.example.com.tflitecamerademo E/AndroidRuntime: FATAL EXCEPTION: CameraBackground
    Process: android.example.com.tflitecamerademo, PID: 22222
    java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: GpuDelegate Prepare: Dimension is empty.Node number 68 (GpuDelegate) failed to prepare.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.applyDelegate(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.init(NativeInterpreterWrapper.java:83)
        at org.tensorflow.lite.NativeInterpreterWrapper.<init>(NativeInterpreterWrapper.java:60)
        at org.tensorflow.lite.Interpreter.<init>(Interpreter.java:224)
        at com.example.android.tflitecamerademo.ImageClassifier.recreateInterpreter(ImageClassifier.java:168)
        at com.example.android.tflitecamerademo.ImageClassifier.useGpu(ImageClassifier.java:176)
        at com.example.android.tflitecamerademo.Camera2BasicFragment.lambda$updateActiveModel$0$Camera2BasicFragment(Camera2BasicFragment.java:379)
        at com.example.android.tflitecamerademo.Camera2BasicFragment$$Lambda$0.run(Unknown Source:8)
        at android.os.Handler.handleCallback(Handler.java:789)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:164)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-01-23 13:51:12.102 4411-8924/? E/CameraDeviceClient: Disconnect from CameraDeviceClient
```

**Comment by ramtin2080** [comment]
Similar problem also exist on some of the MobileNetV1 models (https://tfhub.dev/google/imagenet/mobilenet_v1_050_160/classification/1)
```
2019-01-23 19:59:54.482 5685-5775/android.example.com.tflitecamerademo E/AndroidRuntime: FATAL EXCEPTION: CameraBackground
    Process: android.example.com.tflitecamerademo, PID: 5685
    java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: WARNING: op code #40 cannot be handled by this delegate.  Only the first 29 ops will run on the GPU, and the remaining 4 on the CPU.GpuDelegate Prepare: Dimension is empty.Node number 33 (GpuDelegate) failed to prepare.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.applyDelegate(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.init(NativeInterpreterWrapper.java:83)
        at org.tensorflow.lite.NativeInterpreterWrapper.<init>(NativeInterpreterWrapper.java:60)
        at org.tensorflow.lite.Interpreter.<init>(Interpreter.java:224)
        at com.example.android.tflitecamerademo.ImageClassifier.recreateInterpreter(ImageClassifier.java:168)
        at com.example.android.tflitecamerademo.ImageClassifier.useGpu(ImageClassifier.java:176)
        at com.example.android.tflitecamerademo.Camera2BasicFragment.lambda$updateActiveModel$0$Camera2BasicFragment(Camera2BasicFragment.java:379)
        at com.example.android.tflitecamerademo.Camera2BasicFragment$$Lambda$0.run(Unknown Source:8)
        at android.os.Handler.handleCallback(Handler.java:789)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:164)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-01-23 19:59:54.506 4411-5699/? E/CameraDeviceClient: Disconnect from CameraDeviceClient
2019-01-23 19:59:54.550 4761-4797/? E/PhoneWindow_APM :: sendNotificationForAppPermissionMonitor callingPid = 4761, uid = 1000
2019-01-23 19:59:54.550 4761-4797/? E/PhoneWindow_APM :: sendNotificationForAppPermissionMonitor 2
2019-01-23 19:59:54.550 4761-4797/? E/PhoneWindow_APM :: sendNotificationForAppPermissionMonitor 21000_4761
2019-01-23 19:59:54.550 4761-4797/? E/PhoneWindow_APM :: isCalledPackage return false
2019-01-23 19:59:54.550 4761-4797/? E/PhoneWindow_APM :: sendNotificationForAppPermissionMonitor mIsMonitoredFeature = true
2019-01-23 19:59:54.852 4261-5961/? E/ExynosCameraRequestManager: [CAM_ID(1)][]-ERR(m_resultCallbackThreadFunc[2417]):ResultCallback fail, ret(-110)
2019-01-23 19:59:54.870 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.872 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.872 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.872 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.872 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.890 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.890 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.890 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:54.890 4261-4261/? E/libcsc: csc_deinit:: unsupported csc_hw_type
2019-01-23 19:59:55.258 6499-6676/? E/PBSessionCacheImpl: sessionId[28584452507712951] not persisted.
2019-01-23 19:59:55.440 6365-7849/? E/Launcher.WallpaperUtils: check bitmap.recycle() on releaseWallpaperDrawable
2019-01-23 19:59:57.038 4761-5340/? E/Watchdog: !@Sync 667 [2019-01-23 19:59:57.038]
2019-01-23 19:59:57.326 6189-6473/? E/ContactsProvider_EventLog: Flush buffer to file cnt : 1 size : 1Kb duration : 7ms lastUpdatedAfter : 38673 ms mFlush_time_threasold : 2000 mCurrentSize : 683
2019-01-23 20:00:15.924 7874-7874/? E/Zygote: isWhitelistProcess - Process is Whitelisted
2019-01-23 20:00:15.926 7874-7874/? E/libpersona: scanKnoxPersonas
2019-01-23 20:00:15.926 7874-7874/? E/libpersona: Couldn't open the File - /data/system/users/0/personalist.xml - No such file or directory

```

**Comment by impjdi** [comment]
@ramtin2080 

Thanks for trying out the GPU delegate, and sorry for the late response.  Was swamped with other things last week.  Can you share the two TFLite models?  Doesn't have to have the proper weights; you can share an uninitialized graph.  I want to take a look at the architecture and find out what's wrong.

**Comment by ramtin2080** [comment]
@impjdi

Thank you for the response.  The models are attached.
[tf_files.zip](https://github.com/tensorflow/tensorflow/files/2814332/tf_files.zip)

**Comment by impjdi** [comment]
@ramtin2080 

Perfect.  I'll check these and get back to you by EOD

**Comment by aselle** [comment]
@impjdi, implicit size is not really supported by toco. However, tflite interpreter will not have right shape until a user calls AllocateTensors () (and it will use the input tensors shapes to propagate the shapes downstream), at least for a model with all non-dynamic shapes.

**Comment by impjdi** [comment]
@aselle Thanks! 👍 

@ramtin2080 

The GPU delegate currently does not support dynamic tensor sizes.  Is it possible for you to explicitly specify the tensor dimensions of module_apply_default/hub_input/Mul/y and module_apply_default/hub_input/Sub/y in your TF graph, so that all tensor dimensions are known in advance?

**Comment by ramtin2080** [comment]
@impjdi

Unfortunately I don't know how to do this. Is there any tutorial for it?
I'm kinda new to TensorFlow.

**Comment by ramtin2080** [comment]
@impjdi 

The code can't increase the accuracy like the model normalization ops :(
Is it possible for you guys to add dynamic tensor sizes support for the GPU delegate in the future? or any way so I can simulate the very two first ops in java.
It seems the ops use [this](https://arxiv.org/abs/1502.03167) article for normalizing images.

**Comment by impjdi** [comment]
@ramtin2080 

> The code can't increase the accuracy like the model normalization ops :(

Hm, that's weird.  Maybe we got the parameters wrong, because it should have the same output.

> Is it possible for you guys to add dynamic tensor sizes support for the GPU delegate in the future?

Yes, we will eventually add it, but it comes with bigger API plumbing, and won't be available even if we open source this in near future.  Dynamic tensors may be supported at best in the 2nd half of 2019, and even that would depend on demand... =/

Batch normalization is something different.  With "normalization" I mentioned, is just dividing the input tensor by 255 for every element, but I see that's not easily doable with the current Java APIs.  Let me find other ways.

**Comment by impjdi** [comment]
@ramtin2080 

I just noticed:

> GPU delegate has problem with MobileNetV2. When I select GPU from device list at tflite demo project (https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/java/demo) the app crashes.

Let me try to take a look at these constants.  Maybe we can apply that in the Java layer directly.  Stay tuned.

**Comment by impjdi** [comment]
@ramtin2080 

Aite.  I looked into the constants, and it looks like it's doing 2.0 * X - 1.0 which is remapping the values from 0.0~1.0 to -1.0~1.0.  So let's do that.

> imgData.putFloat(((pixelValue >> 16) & 0xFF) / 255.f * 2.f - 1.f); 
> imgData.putFloat(((pixelValue >> 8) & 0xFF) / 255.f * 2.f - 1.f); 
> imgData.putFloat((pixelValue & 0xFF) / 255.f * 2.f - 1.f); 

Again, that's the code in `ImageClassifierFloatMobileNet.java`.

**Comment by ramtin2080** [comment]
@impjdi 

Thank you so much! Now the accuracy is same as the original MobileNetV2 model =D
Since the problem has been resolved, I'm closing this issue for now.
